### PR TITLE
Prototype Envpool Support

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR adds envpool example. Interestingly, after increasing `num_envs=32`, I was able to solve Pong in 10 mins :D

image

See the tracked experiment in [costa-huang/cleanRL/runs/3rx432mj](https://wandb.ai/costa-huang/cleanRL/runs/3rx432mj)

### CODE REVIEW

File: cleanrl/ppo_atari_envpool.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    return args
+
+
+class RecordEpisodeStatistics(gym.Wrapper):
+    def __init__(self, env, deque_size=100):
+        super(RecordEpisodeStatistics, self).__init__(env)
+        self.num_envs = getattr(env, "num_envs", 1)
+        self.episode_returns = None
+        self.episode_lengths = None
+        self.is_vector_env = True
```
Comment (yooceii) [inline cleanrl/ppo_atari_envpool.py:88]: is_vector_env is not referenced except in line 185 which is a comment.

**Review by yooceii — APPROVED** [review]

**Comment by yooceii** [comment]
Wonder what's the performance comparing with https://github.com/NVlabs/cule with large number of envs.

**Comment by vwxyzjn** [comment]
Ran a hyper-parameter sweep ([sweeps/nfrd091p](https://wandb.ai/costa-huang/cleanRL/sweeps/nfrd091p)) overnight, now i can solve Pong in ~5 mins, according to [runs/opk2dmta](https://wandb.ai/costa-huang/cleanRL/runs/opk2dmta), with hyper parameters

```
--clip-coef=0.2 --num-envs=16 --num-minibatches=8 --num-steps=128 --update-epochs=3
```

D22B5FE2-0515-46FF-91A3-B29B4DF49EBA

**Review by dosssman — APPROVED** [review]
While I am not too familiar with envpool, I got that it is essential related to the environments, and the PPO logic did not seem to have changed.

My attempt converged around 30 minutes, but did use a weaker CPU server than yours, so I suspect the wall time efficiency is highly depended on hardware. Nevertheless, it is still faster than ppo_atari.py which does not use envpool ( using same hyper parameters), which has yet to converge stably enough after 1h30.

For reference, `ppo_atari_envpoo.py` has an SPS of around 1729, while `ppo_atary.py` has an SPS 489.

In any case, this PR looks good for me.
Great work.

**Comment by vwxyzjn** [comment]
E026FB08-DDA8-4B6F-87AD-571B5C935674

https://wandb.ai/vwxyzjn/ppo-details/reports/Envpool--VmlldzoxNDM3ODQz

**Comment by dosssman** [comment]
This time it seems to take around 50 minutes for Pong.

Is the 5 min PPO solving Pong-v5 really due to the hyper parameters mentioned above ?

Also, I noticed that you used the same machine for all the runs, so I was wondering if the concurrence of the training scripts could have some impact on the overall performance too ...

**Comment by vwxyzjn** [comment]
@dosssman it was largely a bit of hyperparameter tuning. Also, I was running these scripts one at a time, so no concurrent issues.

**Pull request merged**